### PR TITLE
📖 Add a warning that only CoreDNS is supported in KCP

### LINKED ISSUE #4026: Document removal of kube-dns support in kubeadm
Kubeadm is going to drop support for kube-dns in version 1.21 (xref https://github.com/kubernetes/kubeadm/issues/1943).

This should not be a problem for CABPK/KCP users, given that almost all of them should be using CoreDNS, however, we should consider if/how to issue a note to our user in case someone is still using kube-dns by twicking KubeadmConfig.Cluster configuration. 

/kind feature

### PR DESCRIPTION
**Pull request by sgirem463**
**What this PR does / why we need it**:
This PR is to address #4026 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
This adds a warning message in the Cluster API Book Section 1.5
Fixes #4026

### CODE REVIEW

**Comment by sgirem463** [comment]
This document change was verified locally by running "make serve-book", the screenshot is:

Screen Shot 2021-03-29 at 11 38 27 PM

**Comment by MarcelMue** [comment]
@sgirem463 CI is currently complaining as you have a reference issue number in the PR title.
PR titles are meant to be purely descriptive of the actual change being done.

So a title such as `📖 Add a warning that only CoreDNS is supported in KCP` would work :)

**Comment by sgirem463** [comment]
@MarcelMue Thanks, title changed as suggested.

File: docs/book/src/tasks/kubeadm-control-plane.md [hunk 1]
```diff
@@ -2,6 +2,14 @@
 
 Using the Kubeadm control plane type to manage a control plane provides several ways to upgrade control plane machines.
 
+<aside class="note warning">
+
+<h1>Warning</h1>
+
+KCP support only CoreDNS as a DNS server; also kube-dns support has been removed by kubeadm in v1.21.
```
Comment (MarcelMue) [inline docs/book/src/tasks/kubeadm-control-plane.md:9]:
```suggestion
KCP supports only CoreDNS as a DNS server; also kube-dns support has been removed by kubeadm in v1.21.
```

Comment (sgirem463) [reply docs/book/src/tasks/kubeadm-control-plane.md]: @MarcelMue updated as suggested, thanks!

**Comment by sgirem463** [comment]
@MarcelMue @JoelSpeed @moshloop Not sure why the tests failed.

**Comment by sbueringer** [comment]
@sgirem463 
it's unrelated
/test pull-cluster-api-test-main

**Comment by MarcelMue** [comment]
@sgirem463 Can you squash the two commits please?

File: docs/book/src/tasks/kubeadm-control-plane.md [hunk 2]
```diff
@@ -2,6 +2,14 @@
 
 Using the Kubeadm control plane type to manage a control plane provides several ways to upgrade control plane machines.
 
+<aside class="note warning">
+
+<h1>Warning</h1>
+
+KCP supports only CoreDNS as a DNS server; also kube-dns support has been removed by kubeadm in v1.21.
```
[the remark this answers is not in the document]
Comment (sgirem463) [reply docs/book/src/tasks/kubeadm-control-plane.md]: Have updated accordingly

**Comment by sgirem463** [comment]
@MarcelMue I had one more change so there are 3 commits to squash. I have never done squash before, not sure if I did it correctly. Logs from my terminal (can you help take a look):

ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git rebase -i HEAD~3
[detached HEAD c5cc8e404] changes related to 'Document removal of kube-dns support in kubeadm #4026'
 Date: Tue Mar 30 00:28:30 2021 -0700
 1 file changed, 8 insertions(+)
Successfully rebased and updated refs/heads/myfeature.
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git status
On branch myfeature
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	docs/book/src/tasks/.kubeadm-control-plane.md.swp

nothing added to commit but untracked files present (use "git add" to track)
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git branch
  master
* myfeature
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git push origin myfeature
To https://github.com/sgirem463/cluster-api.git
 ! [rejected]            myfeature -> myfeature (non-fast-forward)
error: failed to push some refs to 'https://github.com/sgirem463/cluster-api.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git diff
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git pull origin master
From https://github.com/sgirem463/cluster-api
 * branch                master     -> FETCH_HEAD
Already up to date.
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git pull origin myfeature
From https://github.com/sgirem463/cluster-api
 * branch                myfeature  -> FETCH_HEAD
Merge made by the 'recursive' strategy.
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git diff
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git push origin myfeature
Enumerating objects: 2, done.
Counting objects: 100% (2/2), done.
Delta compression using up to 8 threads
Compressing objects: 100% (2/2), done.
Writing objects: 100% (2/2), 538 bytes | 538.00 KiB/s, done.
Total 2 (delta 0), reused 0 (delta 0)
To https://github.com/sgirem463/cluster-api.git
   5cc07a0fe..893b7b560  myfeature -> myfeature
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git branch
  master
* myfeature
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$

**Comment by sgirem463** [comment]
verified locally as:

Screen Shot 2021-03-31 at 1 34 27 AM

**Comment by sbueringer** [comment]
@sgirem463 
The result of the squash is that your PR only contains one commit.

As usual with git there are several ways to achieve that. I usually do it that way:
* `git rebase -i HEAD~5` (5 is the number of commits you want to change), alternatively `git fetch upstream; git rebase -i upstream/master`
* an editor opens and you can decide what you want to change. Usually pick `squash` for all but the newest commit, you can rewrite your commit message if you want with `reword`
* close the editor / save the file
* git push --force your branch

**Comment by MarcelMue** [comment]
You currently have 5 commits now - here is how do it (not saying it's the best way to do it):
```
git checkout myfeature
git reset --soft HEAD~5
git commit -m "Add a warning that only CoreDNS is supported in KCP"
git push --force
```

**Comment by sgirem463** [comment]
@MarcelMue 
I might have messed up the process and branch contents in my previous rebase and push, I am seeing these now:

ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git branch
  master
* myfeature
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git checkout myfeature
Already on 'myfeature'
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git reset --soft HEAD~5
ALLANs-MBP:cluster-api alue$ git commit -m "Add a warning that only CoreDNS is supported in KCP"
[myfeature 8196c51f4] Add a warning that only CoreDNS is supported in KCP
 7 files changed, 85 insertions(+), 55 deletions(-)
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git push --force
fatal: The current branch myfeature has no upstream branch.
To push the current branch and set the remote as upstream, use

    git push --set-upstream origin myfeature

ALLANs-MBP:cluster-api alue$

**Comment by sgirem463** [comment]
@MarcelMue

ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git checkout myfeature
Already on 'myfeature'
ALLANs-MBP:cluster-api alue$ git reset --soft HEAD~5
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git commit -m "Add a warning that only CoreDNS is supported in KCP"
[myfeature 6ce03a6e6] Add a warning that only CoreDNS is supported in KCP
 7 files changed, 85 insertions(+), 55 deletions(-)
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ git push --force
fatal: The current branch myfeature has no upstream branch.
To push the current branch and set the remote as upstream, use

    git push --set-upstream origin myfeature

ALLANs-MBP:cluster-api alue$ 
ALLANs-MBP:cluster-api alue$ 


Should I do "**git push --set-upstream origin myfeature**" first?

**Comment by sgirem463** [comment]
@MarcelMue 
I set my upstream with "git branch -u origin/myfeature"

then repeat your commands, it seems to work now.

**Comment by MarcelMue** [comment]
It looks like you have squashed correctly now but you did reset too many commits. 

See the current changes in this PR.

**Comment by sgirem463** [comment]
@MarcelMue @sbueringer 
Is there a way to fix the current situation?
Or can I abandon this PR and start a new one?

**Comment by MarcelMue** [comment]
> @MarcelMue @sbueringer
> Is there a way to fix the current situation?
> Or can I abandon this PR and start a new one?

I have not been in the situation to be honest.
Can you try: 
`git rebase upstream/master` ?
Ideally `github` should then fix the diff.

**Comment by CecileRobertMichon** [comment]
@sgirem463 thanks for the PR! Make sure you fill out "Fixes #" in the PR description with the issue number this is fixing so it's linked to the issue.

**Comment by sgirem463** [comment]
@CecileRobertMichon Have updated the PR

**Comment by sgirem463** [comment]
@MarcelMue @sbueringer
Have done what sbueringer suggested, looks good now.

**Review by MarcelMue — APPROVED** [review]
Indeed looks good now!

/LGTM

**Review by elmiko — APPROVED** [review]
/lgtm

**Pull request merged**